synapse/lib/cell.py

```diff
@@ -4032,3 +4048,241 @@ async def _runReloadFunc(self, name, func):
+            if s_common.now() > expires:
+                return False, {'mesg': f'API Key is expired: {iden}',
+                               'user': user, 'name': udef.get('name')}
+
+        shadow = kdef.pop('shadow')
+        valid = await s_passwd.checkShadowV2(secv, shadow)  # TODO Timecache this API call
+        if valid is False:
+            return False, {'mesg': f'API Key shadow mismatch: {iden}',
+                           'user': user, 'name': udef.get('name')}
+
+        return True, {'kdef': kdef, 'udef': udef}
+
+    async def modUserApiKey(self, iden, key, valu):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist, cannot modify it: {iden}', iden=iden)
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+
+        vals = {
+            'updated': s_common.now()
+        }
+        if key == 'name':
+            vals['name'] = valu
+        else:
+            raise s_exc.BadArg(mesg=f'Cannot set {key} on user API Keys.', name=key)
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
+
+        logger.info(f'Updated HTTP API Key {iden} for {user.name}, set {key}={valu}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return kdef
+
+    async def regenerateUserApiKey(self, iden, duration=None):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist, cannot regenerated it: {iden}', iden=iden)
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+
+        _, token, shadow = await s_passwd.generateApiKey(iden=iden)
+
+        now = s_common.now()
+        vals = {
+            'shadow': shadow,
+            'updated': now
+        }
+        kdef['shadow'] = shadow
+
+        if duration is None:
+            vals['expires'] = None
+        else:
+            if duration < 1:
+                raise s_exc.BadArg(mesg='Duration must be equal or greater than 1', name='duration')
+            vals['expires'] = now + duration
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
+
+        logger.info(f'Regenerated HTTP API Key {iden} for {user.name}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return token, kdef
+
+    @s_nexus.Pusher.onPush('user:apikey:update')
+    async def _setUserApiKey(self, user, iden, vals):
```

Comment:
Still think this could be unified into `user:apikey:set` and only take `kdef` as an argument. Lets discuss quick :)